framework/src/main/java/org/checkerframework/framework/util/ComplexQualifierHierarchy.java

```diff
@@ -0,0 +1,309 @@
+        }
+        return bottomsMap;
+    }
+
+    /**
+     * Returns the qualifier kind for the given annotation.
+     *
+     * @param anno annotation mirror
+     * @return the qualifier kind for the given annotation
+     */
+    protected QualifierKind getQualifierKind(AnnotationMirror anno) {
+        String name = AnnotationUtils.annotationName(anno);
+        return getQualifierKind(name);
+    }
+
+    /**
+     * Returns the qualifier kind for the annotation with the fully qualified name {@code name}.
+     *
+     * @param name fully qualified annotation name
+     * @return the qualifier kind for the annotation with {@code name}
+     */
+    protected QualifierKind getQualifierKind(String name) {
+        QualifierKind kind = qualifierKindHierarchy.getQualifierKind(name);
+        if (kind == null) {
+            throw new BugInCF("QualifierKind not in hierarchy: %s", name);
+        }
+        return kind;
+    }
+
+    @Override
+    public Set<? extends AnnotationMirror> getTopAnnotations() {
+        return tops;
+    }
+
+    @Override
+    public AnnotationMirror getTopAnnotation(AnnotationMirror start) {
+        QualifierKind kind = getQualifierKind(start);
+        return topsMap.get(kind.getTop());
+    }
+
+    @Override
+    public Set<? extends AnnotationMirror> getBottomAnnotations() {
+        return bottoms;
+    }
+
+    @Override
+    public @Nullable AnnotationMirror getPolymorphicAnnotation(AnnotationMirror start) {
+        QualifierKind polyKind = getQualifierKind(start).getPolymorphic();
+        if (polyKind == null) {
+            return null;
+        }
+        AnnotationMirror poly = kindToAnnotationMirror.get(polyKind);
+        if (poly == null) {
+            throw new TypeSystemError(
+                    "Poly %s has an element. Override ComplexQualifierHierarchy#getPolymorphicAnnotation.",
+                    polyKind);
+        }
+        return poly;
+    }
+
+    @Override
+    public boolean isPolymorphicQualifier(AnnotationMirror qualifier) {
+        return getQualifierKind(qualifier).isPoly();
+    }
+
+    @Override
+    public AnnotationMirror getBottomAnnotation(AnnotationMirror start) {
+        QualifierKind kind = getQualifierKind(start);
+        return bottomsMap.get(kind.getBottom());
+    }
+
+    @Override
+    public boolean isSubtype(AnnotationMirror subAnno, AnnotationMirror superAnno) {
+        QualifierKind subKind = getQualifierKind(subAnno);
+        QualifierKind superKind = getQualifierKind(superAnno);
+        if (subKind.isSubtype(superKind)) {
+            if (superKind.hasElements() && subKind.hasElements()) {
+                return isSubtype(subAnno, subKind, superAnno, superKind);
+            } else {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns whether or not {@code subAnno} is a subtype of {@code superAnno}. Both {@code
+     * subAnno} and {@code superAnno} are annotations with elements.
```

Comment:
I think "are" should be either "at least one has" or "may be".

Reply:
This is the same question as https://github.com/typetools/checker-framework/pull/3404#discussion_r453075154, so I'm going to "resolve" this comment in favor of that one.